# Data Loading & Understanding
Loading the raw Bengaluru Namma Metro (BMRCL) datasets and understanding what we are working with.

We have two datasets sourced from RTI (Right to Information):
- **station-hourly.csv** — hourly boarding (entry) counts per station
- **station-hourly-exits.csv** — hourly exit counts per station

In [1]:
import pandas as pd

In [2]:
df_board = pd.read_csv('data/station-hourly.csv', sep=';')
df_exit = pd.read_csv('data/station-hourly-exits.csv', sep=';')

In [3]:
df_board.head()

,Date,Hour,Station,Ridership
0,2025-08-01,0,Attiguppe,0
1,2025-08-01,1,Attiguppe,0
2,2025-08-01,2,Attiguppe,0
3,2025-08-01,3,Attiguppe,0
4,2025-08-01,4,Attiguppe,6


In [4]:
df_exit.head()

,Date,Hour,Station,Ridership
0,2025-08-01,0,Attiguppe,0
1,2025-08-01,1,Attiguppe,0
2,2025-08-01,2,Attiguppe,0
3,2025-08-01,3,Attiguppe,0
4,2025-08-01,4,Attiguppe,0


In [5]:
df_board.shape

(92280, 4)

In [6]:
df_exit.shape

(95616, 4)

In [7]:
# 3336 more rows in exits, need to figure out why
print(f"Row difference: {df_exit.shape[0] - df_board.shape[0]}")

Row difference: 3336


In [8]:
df_board.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92280 entries, 0 to 92279
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Date       92280 non-null  object
 1   Hour       92280 non-null  int64 
 2   Station    92280 non-null  object
 3   Ridership  92280 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 2.8+ MB


In [9]:
df_exit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95616 entries, 0 to 95615
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Date       95616 non-null  object
 1   Hour       95616 non-null  int64 
 2   Station    95616 non-null  object
 3   Ridership  95616 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 2.9+ MB


In [10]:
df_board.describe()

,Hour,Ridership
count,92280.000000,92280.000000
mean,11.500000,366.687061
std,6.922224,492.533122
min,0.000000,0.000000
25%,5.750000,14.000000
50%,11.500000,213.000000
75%,17.250000,496.000000
max,23.000000,4425.000000


In [11]:
df_exit.describe()

,Hour,Ridership
count,95616.000000,95616.000000
mean,11.500000,352.737000
std,6.922223,529.160717
min,0.000000,0.000000
25%,5.750000,9.000000
50%,11.500000,186.000000
75%,17.250000,456.000000
max,23.000000,9112.000000


In [12]:
df_board.isnull().sum()

Date         0
Hour         0
Station      0
Ridership    0
dtype: int64

In [13]:
df_exit.isnull().sum()

Date         0
Hour         0
Station      0
Ridership    0
dtype: int64

In [14]:
df_board.duplicated().sum()

0

In [15]:
df_exit.duplicated().sum()

0

In [16]:
df_board.dtypes

Date         object
Hour          int64
Station      object
Ridership     int64
dtype: object

## Stations

In [17]:
df_board['Station'].nunique()

83

In [18]:
df_exit['Station'].nunique()

83

In [19]:
# check if both datasets have the same stations
board_stations = set(df_board['Station'].unique())
exit_stations = set(df_exit['Station'].unique())
print(f"Only in Boarding: {board_stations - exit_stations}")
print(f"Only in Exits:    {exit_stations - board_stations}")

Only in Boarding: set()
Only in Exits:    set()


In [23]:
sorted(df_board['Station'].unique())

['Attiguppe',
 'BTM Layout',
 'Baiyappanahalli',
 'Banashankari',
 'Benniganahalli',
 'Beratena Agrahara',
 'Biocon Hebbagodi',
 'Bommanahalli',
 'Central Silk Board',
 'Challaghatta',
 'Chickpete',
 'Chikkabidarakallu',
 'Cubbon Park',
 'Dasarahalli',
 'Deepanjali Nagar',
 'Delta Electronics Bommasandra',
 'Doddakallasandra',
 'Dr. B. R. Ambedkar Station, Vidhana Soudha',
 'Electronic City',
 'Garudacharpalya',
 'Goraguntepalya',
 'Halasuru',
 'Hongasandra',
 'Hoodi',
 'Hopefarm Channasandra',
 'Hosa Road',
 'Huskur Road',
 'Indiranagar',
 'Infosys Foundation Konappana Agrahara',
 'Jalahalli',
 'Jaya Prakash Nagar',
 'Jayadeva Hospital',
 'Jayanagar',
 'Jnanabharathi',
 'Kadugodi Tree Park',
 'Kengeri',
 'Kengeri Bus Terminal',
 'Konanakunte Cross',
 'Krantivira Sangolli Rayanna Railway Station',
 'Krishna Rajendra Market',
 'Krishnarajapura',
 'Kudlu Gate',
 'Kundalahalli',
 'Lalbagh',
 'Madavara',
 'Magadi Road',
 'Mahakavi Kuvempu Road',
 'Mahalakshmi',
 'Mahatma Gandhi Road',
 'Ma

## Date Range

In [24]:
print(f"Boarding: {df_board['Date'].min()} to {df_board['Date'].max()}")
print(f"Exits:    {df_exit['Date'].min()} to {df_exit['Date'].max()}")

Boarding: 2025-08-01 to 2025-09-30
Exits:    2025-08-01 to 2025-09-30


In [25]:
board_dates = sorted(df_board['Date'].unique())
print(f"Total unique dates in boarding: {len(board_dates)}")
board_dates

Total unique dates in boarding: 48


['2025-08-01',
 '2025-08-02',
 '2025-08-03',
 '2025-08-04',
 '2025-08-05',
 '2025-08-06',
 '2025-08-07',
 '2025-08-08',
 '2025-08-09',
 '2025-08-10',
 '2025-08-11',
 '2025-08-12',
 '2025-08-13',
 '2025-08-14',
 '2025-08-15',
 '2025-08-16',
 '2025-08-17',
 '2025-08-18',
 '2025-09-01',
 '2025-09-02',
 '2025-09-03',
 '2025-09-04',
 '2025-09-05',
 '2025-09-06',
 '2025-09-07',
 '2025-09-08',
 '2025-09-09',
 '2025-09-10',
 '2025-09-11',
 '2025-09-12',
 '2025-09-13',
 '2025-09-14',
 '2025-09-15',
 '2025-09-16',
 '2025-09-17',
 '2025-09-18',
 '2025-09-19',
 '2025-09-20',
 '2025-09-21',
 '2025-09-22',
 '2025-09-23',
 '2025-09-24',
 '2025-09-25',
 '2025-09-26',
 '2025-09-27',
 '2025-09-28',
 '2025-09-29',
 '2025-09-30']

In [26]:
exit_dates = sorted(df_exit['Date'].unique())
print(f"Total unique dates in exits: {len(exit_dates)}")
exit_dates

Total unique dates in exits: 48


['2025-08-01',
 '2025-08-02',
 '2025-08-03',
 '2025-08-04',
 '2025-08-05',
 '2025-08-06',
 '2025-08-07',
 '2025-08-08',
 '2025-08-09',
 '2025-08-10',
 '2025-08-11',
 '2025-08-12',
 '2025-08-13',
 '2025-08-14',
 '2025-08-15',
 '2025-08-16',
 '2025-08-17',
 '2025-08-18',
 '2025-09-01',
 '2025-09-02',
 '2025-09-03',
 '2025-09-04',
 '2025-09-05',
 '2025-09-06',
 '2025-09-07',
 '2025-09-08',
 '2025-09-09',
 '2025-09-10',
 '2025-09-11',
 '2025-09-12',
 '2025-09-13',
 '2025-09-14',
 '2025-09-15',
 '2025-09-16',
 '2025-09-17',
 '2025-09-18',
 '2025-09-19',
 '2025-09-20',
 '2025-09-21',
 '2025-09-22',
 '2025-09-23',
 '2025-09-24',
 '2025-09-25',
 '2025-09-26',
 '2025-09-27',
 '2025-09-28',
 '2025-09-29',
 '2025-09-30']

In [27]:
# check if there are dates in one but not the other
print(f"Dates only in exits: {set(exit_dates) - set(board_dates)}")
print(f"Dates only in boarding: {set(board_dates) - set(exit_dates)}")

Dates only in exits: set()
Dates only in boarding: set()


Both datasets cover the same dates and same 83 stations.  
But boarding has fewer rows — lets check which stations have the mismatch.

## Investigating the Row Count Difference

In [28]:
board_counts = df_board.groupby('Station').size().rename('Boarding_Rows')
exit_counts = df_exit.groupby('Station').size().rename('Exit_Rows')

In [29]:
comp = pd.concat([board_counts, exit_counts], axis=1)
comp['Diff'] = comp['Exit_Rows'] - comp['Boarding_Rows']
comp[comp['Diff'] != 0]

,Boarding_Rows,Exit_Rows,Diff
Station,,,
BTM Layout,912,1152,240
Beratena Agrahara,912,1152,240
Biocon Hebbagodi,912,1152,240
Bommanahalli,912,1152,240
Central Silk Board,912,1152,240
Delta Electronics Bommasandra,912,1152,240
Electronic City,912,1152,240
Hongasandra,936,1152,216
Hosa Road,912,1152,240


15 stations have fewer boarding rows than exit rows. All of these are on the Yellow Line 
(R.V. Road to Bommasandra corridor).  

This happens in phased metro rollouts — exit gates record all passengers who deboard, 
but boarding (entry) may be restricted or not logged during certain dates when the line is 
under testing or partial operations.  

We will handle this during preprocessing by doing an outer merge and filling missing boarding counts with 0.

## Hour Values

In [30]:
df_board['Hour'].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23], dtype=int64)

In [31]:
df_board['Hour'].value_counts().sort_index()

Hour
0     3845
1     3845
2     3845
3     3845
4     3845
5     3845
6     3845
7     3845
8     3845
9     3845
10    3845
11    3845
12    3845
13    3845
14    3845
15    3845
16    3845
17    3845
18    3845
19    3845
20    3845
21    3845
22    3845
23    3845
Name: count, dtype: int64

Hours go from 0 to 23 — full 24-hour coverage. Good.

## Quick Ridership Check

In [32]:
print(f"Total passengers boarded: {df_board['Ridership'].sum():,}")
print(f"Total passengers exited:  {df_exit['Ridership'].sum():,}")

Total passengers boarded: 33,837,882
Total passengers exited:  33,727,301


In [33]:
# zero ridership entries — late night hours when metro is closed
print(f"Zero boarding rows: {(df_board['Ridership'] == 0).sum()}")
print(f"Zero exit rows:    {(df_exit['Ridership'] == 0).sum()}")

Zero boarding rows: 18200
Zero exit rows:    21321


In [34]:
# what percentage of rows have zero ridership
print(f"Zero boarding %: {(df_board['Ridership'] == 0).sum() / len(df_board) * 100:.1f}%")

Zero boarding %: 19.7%


## Summary

**What we know so far:**
- Two datasets: boarding (entries) and exits — both from BMRCL Namma Metro, Bengaluru
- 83 stations, hourly granularity, Aug-Sep 2025
- No null values, no duplicates in either dataset
- Boarding has 92,280 rows, Exits has 95,616 rows (3,336 row gap from Yellow Line stations)
- ~33.8 million total passenger boardings recorded
- ~20% of rows are zero (late night hours when metro doesnt run)
- Semicolon delimiter in the raw CSV (not comma) — regional government data formatting